File: engine.py
import random


LETTER_NOT_IN_WORD = "⬜"
CORRECT_LETTER__WRONG_INDEX = "🟨"
CORRECT_LETTER__CORRECT_INDEX = "🟩"


with open("known_words.txt") as file:
  KNOWN_WORDS = file.read().splitlines()


def compare_words(guess: str, secret: str) -> str:
  """
  >>> compare_words("hello", "kelly") == "⬜🟩🟩🟩⬜"
  >>> compare_words("hello", "kohen") == "🟨🟨⬜⬜🟨"
  """
  assert len(guess) == len(secret) == 5
  
  guess = list(guess)
  secret = list(secret)
  
  result = [LETTER_NOT_IN_WORD] * 5
  
  for idx in range(5):
    if guess[idx] == secret[idx]:
      guess[idx] = "_"
      secret[idx] = "_"
      result[idx] = CORRECT_LETTER__CORRECT_INDEX
  
  for idx, letter in enumerate(guess):    
    if letter in secret:
      secret[secret.index(letter)] = "_"
      result[idx] = CORRECT_LETTER__WRONG_INDEX
  
  return "".join(result)


class Game:
  def __init__(self):
    self.__secret_word = random.choice(KNOWN_WORDS)
    self.__guesses_made = 0
  
  
  def make_a_guess(self, guess):
    self.__guesses_made += 1
    return compare_words(guess, self.__secret_word)

  
  def num_guesses_made(self):
    return self.__guesses_made



def __test():
  print("testing the engine...")
  
  game = Game()
  game._Game__secret_word = "hello"
  
  assert game.make_a_guess("fatut") == "⬜⬜⬜⬜⬜"
  assert game.make_a_guess("ehoaa") == "🟨🟨🟨⬜⬜"
  assert game.make_a_guess("helll") == "🟩🟩🟩🟩⬜"
  assert game.make_a_guess("hello") == "🟩🟩🟩🟩🟩"
  
  assert game.num_guesses_made() == 4
  
  print("engine tests passed :)")


if __name__ == "__main__":
  __test()

File: engine.mjs

import fs from 'fs';

export const LETTER_NOT_IN_WORD = "⬜";
export const CORRECT_LETTER_WRONG_INDEX = "🟨";
export const CORRECT_LETTER_CORRECT_INDEX = "🟩";

export const KNOWN_WORDS = fs.readFileSync('known_words.txt', 'utf-8').split(/\s+/g);

export function compareWords(guess, secret) {
  if (guess.length !== 5 || secret.length !== 5) throw new Error('Words must be 5 letters long');

  let guessArr = guess.split('');
  let secretArr = secret.split('');
  let result = Array(5).fill(LETTER_NOT_IN_WORD);

  for (let i = 0; i < 5; i++) {
    if (guessArr[i] === secretArr[i]) {
      guessArr[i] = '_';
      secretArr[i] = '_';
      result[i] = CORRECT_LETTER_CORRECT_INDEX;
    }
  }

  for (let i = 0; i < 5; i++) {
    if (guessArr[i] !== '_' && secretArr.includes(guessArr[i])) {
      secretArr[secretArr.indexOf(guessArr[i])] = '_';
      result[i] = CORRECT_LETTER_WRONG_INDEX;
    }
  }

  return result.join('');
}

export class Game {
  constructor() {
    this.secretWord = KNOWN_WORDS[Math.floor(Math.random() * KNOWN_WORDS.length)];
    this.guessesMade = 0;
  }

  makeAGuess(guess) {
    this.guessesMade += 1;
    return compareWords(guess, this.secretWord);
  }

  numGuessesMade() {
    return this.guessesMade;
  }
}

function __test() {
  console.log("testing the engine...");

  console.assert(compareWords("hello", "kelly") === "⬜🟩🟩🟩⬜");
  console.assert(compareWords("hello", "kohen") === "🟨🟨⬜⬜🟨");

  let game = new Game();
  game.secretWord = "hello";

  console.assert(game.makeAGuess("fatut") === "⬜⬜⬜⬜⬜");
  console.assert(game.makeAGuess("ehoaa") === "🟨🟨🟨⬜⬜");
  console.assert(game.makeAGuess("helll") === "🟩🟩🟩🟩⬜");
  console.assert(game.makeAGuess("hello") === "🟩🟩🟩🟩🟩");

  console.assert(game.numGuessesMade() === 4);

  console.log("engine tests passed :)");
}

__test();
